### PR TITLE
Add secondry deserializer to return amqp envelope

### PR DESCRIPTION
**Pull request by fenryka**
Useful for testing the carpenter by giving access to a constructed
envelope rather than manually building one every time

### CODE REVIEW

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt [hunk 1]
```diff
@@ -57,28 +60,36 @@ class DeserializationInput(internal val serializerFactory: SerializerFactory = S
...
+        if (size.toInt() != bytes.size - 8) {
+            throw NotSerializableException("Unexpected size of data")
+        }
+
+        return Envelope.get(data)
+    }
+
+
+    @Throws(NotSerializableException::class)
+    fun <T : Any, R> des(bytes: SerializedBytes<T>, clazz: Class<T>, generator: (SerializedBytes<T>, Class<T>) -> R): R {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: Should this be private?

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt [hunk 2]
```diff
@@ -88,6 +99,28 @@ class DeserializationInput(internal val serializerFactory: SerializerFactory = S
...
+    @Throws(NotSerializableException::class)
+    fun <T : Any> deserialize(bytes: SerializedBytes<T>, clazz: Class<T>): T {
+        return des<T, T>(bytes, clazz) { bytes, clazz ->
+            var envelope = getEnvelope(bytes)
+            clazz.cast(readObjectOrNull(envelope.obj, envelope.schema, clazz))
+        }
+    }
+
+    @Throws(NotSerializableException::class)
+    fun <T : Any> deserializeAndReturnEnvelope(bytes: SerializedBytes<T>, clazz: Class<T>): classAndEnvelope<T> {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: Could this just be kotlin `internal` which keeps it out of the API but usable within same gradle module?

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt [hunk 3]
```diff
@@ -57,28 +60,36 @@ class DeserializationInput(internal val serializerFactory: SerializerFactory = S
...
-     * be deserialized and a schema describing the types of the objects.
-     */
     @Throws(NotSerializableException::class)
-    fun <T : Any> deserialize(bytes: SerializedBytes<T>, clazz: Class<T>): T {
+    inline fun <reified T : Any> deserializeAndReturnEnvelope(bytes: SerializedBytes<T>): classAndEnvelope<T> =
+            deserializeAndReturnEnvelope(bytes, T::class.java)
+
+
+    @Throws(NotSerializableException::class)
+    fun <T : Any> getEnvelope(bytes: SerializedBytes<T>): Envelope {
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: private or internal?

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt [hunk 4]
```diff
@@ -88,6 +99,28 @@ class DeserializationInput(internal val serializerFactory: SerializerFactory = S
...
+    /**
+     * This is the main entry point for deserialization of AMQP payloads, and expects a byte sequence involving a header
+     * indicating what version of Corda serialization was used, followed by an [Envelope] which carries the object to
+     * be deserialized and a schema describing the types of the objects.
+     */
+    @Throws(NotSerializableException::class)
+    fun <T : Any> deserialize(bytes: SerializedBytes<T>, clazz: Class<T>): T {
+        return des<T, T>(bytes, clazz) { bytes, clazz ->
+            var envelope = getEnvelope(bytes)
+            clazz.cast(readObjectOrNull(envelope.obj, envelope.schema, clazz))
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: BTW, I have ambitions that we can later make the envelope extraction lazy, so we only do all the parsing if it turns out we haven't seen all the signatures/descriptors before.  The AMQP stuff is length encoded.  I've succeeded in peeking inside for something I'm working on so will definitely be possible.  But for now this is fine...

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt (hunk 1 above)
Comment (fenryka) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: most probably... done!

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt (hunk 3 above)
Comment (fenryka) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: done!

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt (hunk 2 above)
Comment (fenryka) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: done!

File: core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt [hunk 5]
```diff
@@ -11,6 +11,9 @@ import java.lang.reflect.Type
 import java.nio.ByteBuffer
 import java.util.*
 
+
+data class classAndEnvelope<T>(val obj: T, val envelope: Envelope)
```
Comment (fenryka) [inline core/src/main/kotlin/net/corda/core/serialization/amqp/DeserializationInput.kt]: Done!!!!

**Comment by fenryka** [comment]
*pokes* @rick-r3 for second review with changes :)

File: core/src/test/kotlin/net/corda/core/serialization/amqp/DeserializeAndReturnEnvelope.kt [hunk 6]
```diff
@@ -0,0 +1,42 @@
+package net.corda.core.serialization.amqp
+
+import org.junit.Test
+import kotlin.test.*
+
+class DeserializeAndReturnEnvelope {
```
Comment (rick-r3) [inline core/src/test/kotlin/net/corda/core/serialization/amqp/DeserializeAndReturnEnvelope.kt]: Generally we've appended `Tests` or `Test` to the name of classes containing tests, so when code completing in the IDE it's obvious.

**Pull request merged**